File: 백준/Gold/7662. 이중 우선순위 큐/이중 우선순위 큐.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const T = parseInt(input[0]);
let idx = 1;

class Heap {
  constructor(comp) {
    this.data = [];
    this.comp = comp;
  }

  push(val) {
    this.data.push(val);
    this._heapifyUp();
  }

  pop() {
    if (this.size() === 0) return null;
    const top = this.data[0];
    const end = this.data.pop();
    if (this.size() > 0) {
      this.data[0] = end;
      this._heapifyDown();
    }
    return top;
  }

  peek() {
    return this.data[0] ?? null;
  }

  size() {
    return this.data.length;
  }

  _heapifyUp() {
    let i = this.size() - 1;
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (this.comp(this.data[i], this.data[p])) {
        [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
        i = p;
      } else break;
    }
  }

  _heapifyDown() {
    let i = 0;
    const n = this.size();
    while (2 * i + 1 < n) {
      let left = 2 * i + 1;
      let right = 2 * i + 2;
      let target = left;

      if (right < n && this.comp(this.data[right], this.data[left])) {
        target = right;
      }

      if (this.comp(this.data[target], this.data[i])) {
        [this.data[i], this.data[target]] = [this.data[target], this.data[i]];
        i = target;
      } else break;
    }
  }
}

for (let t = 0; t < T; t++) {
  const k = parseInt(input[idx++]);
  const minHeap = new Heap((a, b) => a < b);
  const maxHeap = new Heap((a, b) => a > b);
  const countMap = new Map();

  for (let i = 0; i < k; i++) {
    const [cmd, valStr] = input[idx++].split(' ');
    const val = parseInt(valStr);

    if (cmd === 'I') {
      minHeap.push(val);
      maxHeap.push(val);
      countMap.set(val, (countMap.get(val) || 0) + 1);
    } else {
      const targetHeap = val === 1 ? maxHeap : minHeap;

      while (targetHeap.size()) {
        const num = targetHeap.pop();
        if (countMap.get(num)) {
          countMap.set(num, countMap.get(num) - 1);
          break;
        }
      }
    }
  }

  let max = null;
  let min = null;

  while (maxHeap.size()) {
    const num = maxHeap.pop();
    if (countMap.get(num)) {
      max = num;
      break;
    }
  }

  while (minHeap.size()) {
    const num = minHeap.pop();
    if (countMap.get(num)) {
      min = num;
      break;
    }
  }

  if (max === null || min === null) {
    console.log('EMPTY');
  } else {
    console.log(`${max} ${min}`);
  }
}

File: 백준/Gold/7662. 이중 우선순위 큐/이중 우선순위 큐.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// ---------- 일반화된 Heap 구조체 ----------
type Heap struct {
	data []int
	less func(a, b int) bool
}

func (h Heap) Len() int           { return len(h.data) }
func (h Heap) Less(i, j int) bool { return h.less(h.data[i], h.data[j]) }
func (h Heap) Swap(i, j int)      { h.data[i], h.data[j] = h.data[j], h.data[i] }

func (h *Heap) Push(x any) {
	h.data = append(h.data, x.(int))
}

func (h *Heap) Pop() any {
	n := len(h.data)
	val := h.data[n-1]
	h.data = h.data[:n-1]
	return val
}

func (h *Heap) Peek() (int, bool) {
	if len(h.data) == 0 {
		return 0, false
	}
	return h.data[0], true
}

func (h *Heap) Empty() bool {
	return len(h.data) == 0
}

// ---------- 입력 처리 ----------
var (
	s = bufio.NewScanner(os.Stdin)
	w = bufio.NewWriter(os.Stdout)
)

func main() {
	defer w.Flush()
	s.Split(bufio.ScanLines)

	T := scanInt()
	for t := 0; t < T; t++ {
		k := scanInt()

		minHeap := &Heap{less: func(a, b int) bool { return a < b }}
		maxHeap := &Heap{less: func(a, b int) bool { return a > b }}
		heap.Init(minHeap)
		heap.Init(maxHeap)
		countMap := make(map[int]int)

		for i := 0; i < k; i++ {
			parts := scanWords()
			cmd := parts[0]
			num, _ := strconv.Atoi(parts[1])

			if cmd == "I" {
				heap.Push(minHeap, num)
				heap.Push(maxHeap, num)
				countMap[num]++
			} else {
				if num == 1 {
					// 최대값 삭제
					for !maxHeap.Empty() {
						v := heap.Pop(maxHeap).(int)
						if countMap[v] > 0 {
							countMap[v]--
							break
						}
					}
				} else {
					// 최소값 삭제
					for !minHeap.Empty() {
						v := heap.Pop(minHeap).(int)
						if countMap[v] > 0 {
							countMap[v]--
							break
						}
					}
				}
			}
		}

		// 유효한 최대값 추출
		var maxVal *int = nil
		for !maxHeap.Empty() {
			v := heap.Pop(maxHeap).(int)
			if countMap[v] > 0 {
				maxVal = &v
				break
			}
		}

		// 유효한 최소값 추출
		var minVal *int = nil
		for !minHeap.Empty() {
			v := heap.Pop(minHeap).(int)
			if countMap[v] > 0 {
				minVal = &v
				break
			}
		}

		if maxVal == nil || minVal == nil {
			fmt.Fprintln(w, "EMPTY")
		} else {
			fmt.Fprintf(w, "%d %d\n", *maxVal, *minVal)
		}
	}
}

// ---------- 유틸 ----------
func scanWords() []string {
	s.Scan()
	return strings.Fields(s.Text())
}

func scanInt() int {
	s.Scan()
	v, _ := strconv.Atoi(s.Text())
	return v
}
